fairscale/experimental/nn/data_parallel/gossip/distributed.py

```diff
@@ -0,0 +1,1192 @@
+            ):
+                for p_fp16, p_fp32 in zip(p_fp16_group, p_fp32_group):
+                    yield p_fp16, p_fp32
+
+        # Handle fairseq fp16 optimizer
+        elif fp32_params is not None:
+            if isinstance(fp32_params, dict):
+                fp32_params_list = list(fp32_params.values())
+                assert len(fp32_params_list) == 1
+                fp32_params = fp32_params_list[0]
+
+            if isinstance(fp32_params, list):
+                for p, fp32_param in zip(self.parameters(), fp32_params):
+                    yield p.view(-1), fp32_param
+            else:
+                offset = 0
+                for p in self.parameters():
+                    yield p.view(-1), fp32_params[offset : offset + p.numel()]
+                    offset += p.numel()
+
+    def _should_perform_slowmo(self) -> bool:
+        return self.slowmo and (self.num_updates + 1) % self.slowmo_frequency == 0
+
+    def _should_perform_localsgd(self) -> bool:
+        return self.localsgd and (self.num_updates + 1) % self.localsgd_frequency == 0
+
+    def _skip_averaging_memory_efficient_slowmo(self) -> bool:
+        return self.slowmo_memory_efficient and self._should_perform_slowmo()
+
+    def _should_perform_sgp_common(self) -> bool:
+        return self.sgp and not self.overlap and not self._skip_averaging_memory_efficient_slowmo()
+
+    def _should_perform_sgp(self) -> bool:
+        return self._should_perform_sgp_common() and not self.overlap
+
+    def _should_perform_sgp_overlap(self) -> bool:
+        return self._should_perform_sgp_common() and self.overlap
+
+    def _should_use_error_feedback(self, fp16_fp32_list: List[Tuple[torch.Tensor, torch.Tensor]]) -> bool:
+        return bool(fp16_fp32_list) and (self._should_perform_sgp() or self._should_allreduce_params())
+
+    def _should_allreduce_params(self) -> bool:
+        # We do not all-reduce parameters with local SGD if a slow momentum step is
+        # performed, since this step contains a reduce operation already. Note that this
+        # also means there is no error feedback correction in that case: it is not needed
+        # since communication within the slow momentum step happens in fp32.
+        return (self.sgp and self._should_perform_slowmo() and self.slowmo_sgp_average_params) or (
+            self._should_perform_localsgd() and not self._skip_averaging_memory_efficient_slowmo()
+        )
+
+    def _maybe_pre_communicate_error_feedback(self, fp16_fp32_list: List[Tuple[torch.Tensor, torch.Tensor]]) -> None:
+        ef_rec = self._create_event_recorder("Error feedback")
+        if self._should_use_error_feedback(fp16_fp32_list):
+            with torch.no_grad():
+                for p_fp16, p_fp32 in fp16_fp32_list:
+                    if self._should_allreduce_params():
+                        # This division and multiplication with the same number is done
```

Comment:
it might be me, but I'm lacking some context on what "error feedback" means here. Something like "storing the lost information in error feedback" depends on some priori knowledge of the field, or at least I hope so ?

Reply:
Yes, indeed it depends on priori knowledge of the field. I have added a link to the original paper to help improve documentation in case someone might be interested in learning about it